Framework/PythonInterface/plugins/algorithms/Abins2D.py

```diff
@@ -0,0 +1,291 @@
+        prog_reporter.resetNumSteps(1, 0.8, 0.80000001)
+        prog_reporter.report("Dynamical structure factors have been determined.")
+        # Now determine number of remaining messages and set reporter for rest of run:
+        n_messages = 3 + bool(self._sum_contributions) + bool(self._save_ascii)
+        prog_reporter.resetNumSteps(n_messages, 0.8, 1)
+
+        # 4) get atoms for which S should be plotted
+        atoms_data = ab_initio_data.get_atoms_data()
+        atom_numbers, atom_symbols = self.get_atom_selection(atoms_data=atoms_data, selection=self._atoms)
+        prog_reporter.report("Atoms, for which dynamical structure factors should be plotted, have been determined.")
+
+        # 5) create workspaces for atoms in interest
+        self.setProperty('OutputWorkspace', self._out_ws_name)
+
+        workspaces = []
+        workspaces.extend(self.create_workspaces(atoms_symbols=atom_symbols, s_data=s_data, atoms_data=atoms_data,
+                                                 max_quantum_order=self._max_event_order))
+        workspaces.extend(self.create_workspaces(atom_numbers=atom_numbers, s_data=s_data, atoms_data=atoms_data,
+                                                 max_quantum_order=self._max_event_order))
+        prog_reporter.report("Workspaces with partial dynamical structure factors have been constructed.")
+
+        # 6) Create a workspace with sum of all atoms if required
+        if self._sum_contributions:
+            self.create_total_workspace(workspaces)
+            prog_reporter.report("Workspace with total S has been constructed.")
+
+        GroupWorkspaces(InputWorkspaces=workspaces, OutputWorkspace=self._out_ws_name)
+
+        # 8) save workspaces to ascii_file
+        if self._save_ascii:
+            self.write_workspaces_to_ascii(ws_name=self._out_ws_name,
+                                           scale=(1.0 / self._instrument.get_energy_bin_width()))
+            prog_reporter.report("All workspaces have been saved to ASCII files.")
+
+        # 9) set  OutputWorkspace
+        self.setProperty('OutputWorkspace', self._out_ws_name)
+        prog_reporter.report("Group workspace with all required  dynamical structure factors has been constructed.")
+
+    def _fill_s_workspace(self, s_points=None, workspace=None, protons_number=None, nucleons_number=None):
+        """
+        Puts S into workspace(s).
+
+        :param s_points: dynamical factor for the given atom
+        :param workspace:  workspace to be filled with S
+        :param protons_number: number of protons in the given type fo atom
+        :param nucleons_number: number of nucleons in the given type of atom
+        """
+        from abins.constants import FUNDAMENTALS, TWO_DIMENSIONAL_INSTRUMENTS
+
+        if self._instrument.get_name() not in TWO_DIMENSIONAL_INSTRUMENTS:
+            raise ValueError("Instrument {self._instrument_name} is not supported by this version of Abins")
+
```

Comment:
```suggestion
elif protons_number > nucleons_number:
    raise ValueError("The proton number is greater than the nucleon number")
```
or something similar

Reply:
If this is going to be checked, it would make more sense to put it in the AtomsData class so it is caught as soon as the data is loaded. I think this can be a separate issue?

Reply:
Issue created
https://github.com/mantidproject/mantid/issues/34004